Hw 11

AntonKutl opened the pull request.

AntonKutl committed:
Добавленна заготовка из задания

AntonKutl committed:
Добавленна заготовка из задания

AntonKutl committed:
задание выполнено

petrelevich commented on hw11-Cache/src/main/java/DbServise/dao/SlowDataSrc.java:
```diff
@@ -0,0 +1,41 @@
+ * created on 14.12.18.
+ */
+public class SlowDataSrc<K, V> {
+    List<WeakReference<HwListener>> listListener = new ArrayList<>();
```
тут и Map<K, V> mapCache = new WeakHashMap<>();
должны быть private final

petrelevich commented on hw11-Cache/src/main/java/DbServise/dao/SlowDataSrc.java:
```diff
@@ -0,0 +1,41 @@
+    private static final Logger logger = LoggerFactory.getLogger(HWCacheDemo.class);
+
+    public void save(K key, V value) {
+        mapCache.put((K) new WeakReference<>(key), value);
```
тут так должно быть mapCache.put(key, value);

petrelevich commented on hw11-Cache/src/main/java/str/Cache.java:
```diff
@@ -0,0 +1,43 @@
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import ru.otus.references.WeakMapDemo;
```
С этим import-ом не компилируется.

AntonKutl committed:
Delete Cache.java

AntonKutl committed:
Delete DataProcessor.java

AntonKutl committed:
Delete Data.java

AntonKutl committed:
Delete WeakMapDemo.java

AntonKutl committed:
Delete ReferenceDemo.java

AntonKutl committed:
задание выполнено

AntonKutl committed:
Merge remote-tracking branch 'origin/hw-11' into hw-11

AntonKutl committed:
задание выполнено

AntonKutl committed:
задание выполнено

AntonKutl committed:
задание выполнено

petrelevich commented on hw11-Cache/pom.xml:
```diff
@@ -8,5 +8,52 @@
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-compiler-plugin</artifactId>
+                <configuration>
+                    <source>8</source>
```
имеет смысл обновиться на java 11

petrelevich commented on hw11-Cache/src/main/java/Main.java:
```diff
@@ -0,0 +1,28 @@
+        }
+
+        for (int i = 1; i <100 ; i++) {
+            ((DAOUserImpl) daoUserUser).readCache(i,User.class);
```
какую роль выполняет этот метод?
только для тестирования?

petrelevich commented on hw11-Cache/src/main/java/dao/DAOUserImpl.java:
```diff
@@ -0,0 +1,69 @@
+    }
+
+    @Override
+    public <T> T read(long id, Class<T> clazz) {
```
Этот метод сначала должен обратиться в кеш, если ничего не нашел, то идти в базу данных.

petrelevich commented on hw11-Cache/src/main/java/cachehw/MyCache.java:
```diff
@@ -0,0 +1,64 @@
+    private List<WeakReference<HwListener>> listListener = new ArrayList<>();
+    private Map<K, V> mapCache = new WeakHashMap<>();
+    public void put(K key, V value) {
+        mapCache.put((K) new WeakReference<>(key), value);
```
WeakReference сделает WeakHashMap, вам это делать не надо.

petrelevich commented on hw11-Cache/src/main/java/dao/DAOUserImpl.java:
```diff
@@ -0,0 +1,69 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class DAOUserImpl<T> implements DAOUser<T> {
```
Сделать свое DAO с кешом - решение правильное.

AntonKutl committed:
Исправил замечания.
mapCache.put((K) new WeakReference<>(key), value) - не мог отцепить сильные ссылки от int

petrelevich commented on hw11-Cache/src/test/java/TestCache.java:
```diff
@@ -0,0 +1,30 @@
+                (key, value, action) -> logger.info("key:{}, value:{}, action: {}",  key, value, action);
+        cache.addListener(listener);
+        for (int i = 0; i <100; i++) {
+            Integer temp=new Integer(i);
```
Обратите внимание, конструктор @Deprecated(since="9")

AntonKutl replied:
Да, но если использовать Integer.valueOf(int) значение в мапе сбрасываться не будут при вызове gc

petrelevich replied:
В новом коде Depricated ни при каких обстоятельствах не стоит использовать.
Для решения проблемы кеширования переходите на String.

petrelevich commented on hw11-Cache/src/test/java/TestCacheBD.java:
```diff
@@ -0,0 +1,41 @@
+
+
+@DisplayName("MyCache должен")
+public class TestCacheBD {
```
не забывайте, что все классы должны быть в пакетах.

petrelevich commented on hw11-Cache/src/main/java/cachehw/MyCache.java:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    public boolean isKey(K key){
+        Set<K> set=  mapCache.keySet();
```
почему не mapCache.keySet().contains()?

petrelevich commented on hw11-Cache/src/main/java/cachehw/MyCache.java:
```diff
@@ -0,0 +1,51 @@
+    private final Map<K, V> mapCache = new WeakHashMap<>();
+    public void put(K key, V value) {
+        mapCache.put(key, value);
+        listListener.stream().forEach((s) -> s.get().notify(key, value, "put"));
```
тут есть возможность NPE получить.

AntonKutl replied:
к сожалению не могу понять как

petrelevich replied:
Посмотрите JavaDoc на метод s.get().
Подумайте, какие значения и в каких случаях может вернуть этот метод, и что с этим можно сделать.

petrelevich commented on hw11-Cache/src/main/java/dao/DAOUserCacheImpl.java:
```diff
@@ -0,0 +1,78 @@
+public class DAOUserCacheImpl<T> implements DAOUser<T> {
+    private SessionFactory sessionFactory;
+    private StandardServiceRegistry serviceRegistry;
+    private MyCache myCache = new MyCache();
```
про final не забывайте и про generics.

petrelevich commented on hw11-Cache/src/main/java/dao/DAOUserCacheImpl.java:
```diff
@@ -0,0 +1,78 @@
+
+        if (myCache.isKey(id)) {
+            logger.info("Четение из кэша ключа:{}, и значения:{}", id, myCache.get(id));
+            return (T) myCache.get(id);
```
тут совсем не обязательно второй раз из кеша значение читать.

petrelevich commented on hw11-Cache/src/main/java/dao/DAOUserCacheImpl.java:
```diff
@@ -0,0 +1,78 @@
+    }
+
+    //метод для теста
+    public <T> T readBD(long id, Class<T> clazz) {
```
придумайте как протестировать работу с кешем без этого метода.

petrelevich commented on hw11-Cache/src/main/java/cachehw/HWCacheDemo.java:
```diff
@@ -0,0 +1,31 @@
+        System.gc();
+        Thread.sleep(1000);
+
+        System.out.println(((MyCache<Integer, Integer>) cache).size());
```
у вас есть logger, пользуйтесь им.

AntonKutl committed:
Исправил замечания.

AntonKutl committed:
Решение по NPE не нашел. В функцию кидал различные значения параметров ошибки не вылазило

petrelevich commented on hw11-Cache/src/main/java/cachehw/MyCache.java:
```diff
@@ -0,0 +1,45 @@
+    private final Map<K, V> mapCache = new WeakHashMap<>();
+    public void put(K key, V value) {
+        mapCache.put(key, value);
+        listListener.stream().forEach((s) -> (Objects.requireNonNull(s.get())).notify(key, value, "put"));
```
функция Objects.requireNonNull:
    public static <T> T requireNonNull(T obj) {
        if (obj == null)
            throw new NullPointerException();
        return obj;
    }
выбросит исключение, если WeakReference<HwListener<K,V>>> listListener протухнет.
при том, что "протухание" - это нормальное явление.
Тут надо просто проверить на null.
Если объект null, то просто идем дальше.

AntonKutl committed:
+

petrelevich commented on hw11-Cache/src/main/java/cachehw/MyCache.java:
```diff
@@ -0,0 +1,57 @@
+    private final Map<K, V> mapCache = new WeakHashMap<>();
+    public void put(K key, V value) {
+        mapCache.put(key, value);
+        for (WeakReference<HwListener<K,V>> temp:listListener) {
```
название переменной не очень удачное.
